Add content for Quickstarts guides

dfennessy opened the pull request:
Done for [THREESCALE-8187](https://issues.redhat.com/browse/THREESCALE-8187)

junezhang commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -0,0 +1,133 @@
+# Additional info: https://docs.openshift.com/container-platform/4.9/web_console/creating-quick-start-tutorials.html
```
IMO, the content PR should only include this one file, we can change the name to basicAPIintegration.yaml.
And the other files will be duplicated with [PR#2903](https://github.com/3scale/porta/pull/2903)
So this is the reason why I suggest creating content PR after PR#2903 has been merged?
cc @josemigallas here.

dfennessy replied:
Is it ok for me to rename the file without consequences elsewhere?

Let me know @josemigallas 

Cheers.

dfennessy replied:
@josemigallas I was wondering about June's question here about changing the file name?

junezhang replied:
I think in the quick start page, we will have eight cards to show different quick start content, such as Basic API Integration, create backend, create product...... 
then here we will have eight YAML files for every content.
Yes, if you change the file name to basicAPIintegraton.yaml, then we need to update the index.js to include the new files.

dfennessy replied:
Thank you for this information. What are the standard for file naming? For example: `basicAPIintegraton.yaml` or `basicapiintegraton.yaml` or `basic-api-integraton.yaml?

Let's discuss in our stand-up.

junezhang commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -0,0 +1,133 @@
+  # you can add additional metadata here
+  instructional: true
+spec:
+  displayName: Complete a basic API Integration
```
displayName should be "Complete a basic API integration"

junezhang commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -0,0 +1,133 @@
+  durationMinutes: 10
+  # Optional type section, will display as a tile on the card
+  type: 
+    text: Type
```
Do you think we use "Quick Start" will be better? @deernano
Keep consistent with other applications.
![截屏2022-02-23 下午3 39 41](https://user-images.githubusercontent.com/701009/155278211-6dfa89a7-05e3-4fb9-b42d-4d1b1f6508d2.png)

junezhang commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -0,0 +1,133 @@
+  description: |-
+    Create the required objects to complete a basic integration with 3scale.
+  introduction: |-
+    This quick start guides you to create the required objects to complete a basic integration with 3scale as an API provider including 7 tasks. To learn more about API configuration, see the [3scale documentation](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.11/html/getting_started/first-steps-with-threescale_configuring-your-api#products-backends-for-threescale- ...
```
About the link, should it be a dynamic link? not a hard code. @josemigallas

dfennessy replied:
@jschuler Maybe you can be of assistance regarding dynamic links in YAML files. Cheers.

dfennessy replied:
Leaving this one until @jschuler replies.

junezhang commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -0,0 +1,133 @@
+  tasks:
+    - title: Create a new backend
+      description: |-
+        ## Text
```
For this piece of code, it is not related to "Create a new backend", check the screenshot.
It's not our real content, it's the demo content.
![截屏2022-02-23 下午3 45 55](https://user-images.githubusercontent.com/701009/155278962-eb97372d-662a-4ebd-b10a-5c0cbc2abbef.png)

junezhang reviewed (changes requested):
Hi, Darren, I just reviewed the templated.yaml file, but I haven't found the real content after the seven titles, when we click the sub-title, we only see the demo content like the screenshot of what paste above.

dfennessy commented:
@junezhang @deernano @josemigallas 

I've added the content, however I'm unsure how to add the tiles for this content via the YAML file.

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -18,58 +18,78 @@ spec:
+        1. Go to _Backends_ page: click the context selector on the header and select <strong>Backends</strong>.
+        1. On the _Backends_ page, click <strong>Create backend</strong>.
+        1. On the _New backend_ page, fill in the details to create the backend.
+        1. Click <strong>Create backend</strong>.
```
Last time we have talked to not having "." after button name?

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -18,58 +18,78 @@ spec:
+        An API product is a bundle of one or more API backends. For this quick start, first generate an API product and name it before bundling backends to this product.
+
+        ### Note
+        Important fields:
```
The Important fields section is under Procedure No.3

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -18,58 +18,78 @@ spec:
+        1. Go to _Products_ page: click the context selector on the header and select <strong>Products</strong>.
+        1. On the _Products_ page, click <strong>Create product</strong>.
+        1. On the _New Product_ page, fill in the details to create the product.
+        1. Click <strong>Create product</strong>.
```
Remove the '."

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -18,58 +18,78 @@ spec:
 
-        The syntax for multi-line code snippets:
-        - Text between triple back quotes, followed by `{{copy}}`
+        ### Note
```
The Important fields section is under Procedure No.3

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -18,58 +18,78 @@ spec:
+        1. On <strong>[Your\_product\_name]</strong> overview page, navigate to <strong>Integration > Backends</strong>.
+        1. Click <strong>Add backend</strong>.
+        1. On the _Add a backend_ page, fill in the details.
+        1. Click <strong>Add to product</strong>.
```
Remove the '."

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -80,20 +100,102 @@ spec:
+        An application plan defines the rules such as limits, pricing, and features for using your API product.
+
+        ### Note
+        Important fields:
```
The Important fields section is under Procedure No.3

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -80,20 +100,102 @@ spec:
+        1. On <strong>[Your\_product\_name]</strong> overview page.
+        1. Click <strong>Create application plan</strong>.
+        1. On the _New application plan_ page, fill in the details.
+        1. Click <strong>Create application plan</strong>.
```
Remove the '."

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -80,20 +100,102 @@ spec:
+
+        ### Note
+        Important fields:
+        - Application plan: An application is always associated with an application plan.
```
The Important fields section is under Procedure No.4

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -80,20 +100,102 @@ spec:
+        1. On the _Publish application plans_ card, click the <strong>x live applications</strong> link.
+        1. On the Applications on your <strong>[Your\_product\_name]</strong> page, click <strong>Create application</strong>.
+        1. On the _New application page_, fill in the details.
+        1. Click <strong>Create application</strong>.
```
Remove the '."

deernano commented on app/javascript/src/QuickStarts/templates/template.yaml:
```diff
@@ -80,20 +100,102 @@ spec:
+      # optional - the task's Check your work module
+      review:
+        instructions: |-
+          - Did you complete the task successfully?
```
change to: 

- Did you create an application plan for your product?
- Did you create an application that subscribes to the application plan?
-Did you promote the APIcast configuration to staging and production?

josemigallas commented:
@dfennessy I'm afraid dynamic content isn't possible in YAML. To have it, we'd need to dynamically generate the whole templates from our backend, which would require to implement some logic in Ruby and would make it harder to maintain in the end.
I say if the amount of links is manageable, let's go with hardcoded.

junezhang commented:
> @dfennessy I'm afraid dynamic content isn't possible in YAML. To have it, we'd need to dynamically generate the whole templates from our backend, which would require to implement some logic in Ruby and would make it harder to maintain in the end. I say if the amount of links is manageable, let's go with hardcoded.

If it's hard for the dynamic content, @deernano how do you think we switch to a static link such as the 3scale overview page in the customer portal https://access.redhat.com/products/red-hat-3scale not a version link?

dfennessy commented:
> <img alt="屏幕快照 2022-03-01 下午3 25 03" width="399" src="https://user-images.githubusercontent.com/5396309/156123783-760c4b26-a68c-483b-b3a2-7cd574761e66.png">
> 
> seems that the ' -' (path) and '1' (4. click to Add to product）are not activated

This looks ok to me at the moment. Maybe it was fixed when I move the `Important fields:` and indented it properly.

dfennessy commented:
> @junezhang @deernano @josemigallas
> 
> I've added the content, however I'm unsure how to add the tiles for this content via the YAML file.

This needs to be addressed, I think. Any ideas?

junezhang commented:
> > @junezhang @deernano @josemigallas
> > I've added the content, however I'm unsure how to add the tiles for this content via the YAML file.
> 
> This needs to be addressed, I think. Any ideas?

To be honest, I didn't understand what's your meaning of the tiles, we can talk about it in our stand-up meeting.

dfennessy commented:
> > > @junezhang @deernano @josemigallas
> > > I've added the content, however I'm unsure how to add the tiles for this content via the YAML file.
> > 
> > 
> > This needs to be addressed, I think. Any ideas?
> 
> To be honest, I didn't understand what's your meaning of the tiles, we can talk about it in our stand-up meeting.

For example, I've added all the content for `template.yaml` under **Complete a basic API integration**. Only one tile appears. We want 8 tiles if there's a separate tile for each....

junezhang commented on app/javascript/src/QuickStarts/templates/index.js:
```diff
@@ -1,6 +1,6 @@
 // @flow
 
-import template from './template.yaml'
+import template from './basicAPIintegration.yaml'
```
rename template to basicAPIintegration

junezhang commented on app/javascript/src/QuickStarts/templates/index.js:
```diff
@@ -1,6 +1,6 @@
+import template from './basicAPIintegration.yaml'
 
 const resources = [
     template
```
add a new resource - basicAPIintegration

deernano commented on app/javascript/src/QuickStarts/templates/basicAPIintegration.yaml:
```diff
@@ -0,0 +1,200 @@
+        
+          Important fields:
+              - _System name_:  3scale needs this name to internally identify and refer to the backend.
+              - _Private Base URL_: It is the base URL endpoint of the private API.
```
It might be better to have one line between No.40 and N0.41


<img width="405" alt="屏幕快照 2022-03-14 上午11 17 12" src="https://user-images.githubusercontent.com/5396309/158099023-ffda60de-9d95-4b68-892f-b3f58627372d.png">

dfennessy replied:
@josemigallas @deernano I'm pretty sure this is a css stylesheet solution. I'm not sure where to place the style and if it'll mess other things up because it's probably a global solution for nested unordered lists. Something like the following:

`ul.frames {
  font-size: 0;
}
ul.frames li {
  font-size: 14px; font-size:1.4rem;
  display: inline;
}`

deernano replied:
@dfennessy I see.  Let's just leave as what it is for now. Thx!

deernano commented on app/javascript/src/QuickStarts/templates/basicAPIintegration.yaml:
```diff
@@ -0,0 +1,200 @@
+        
+          Important fields:
+              - _System name_:  3scale needs this name to internally identify and refer to the backend.
+                - A product _system\_name_ is used to generate proxy endpoints and domain names.
```
It might be better to have one line between No.64 and N0.65

deernano commented on app/javascript/src/QuickStarts/templates/basicAPIintegration.yaml:
```diff
@@ -0,0 +1,200 @@
+              - _Path_: when you add a backend to a product, you define the path that the backend uses to communicate with the specified product. When defining the path for a backend:
+                - You can indicate / as the path of one of the backends.
+                - Paths must be unique inside the product.
+                - You can give the same backend the same path in different products.
```
It might be better to have one line between No.90 and No.91

deernano commented on app/javascript/src/QuickStarts/templates/basicAPIintegration.yaml:
```diff
@@ -0,0 +1,200 @@
+          Important fields:
+              - _System name_: A system name must be unique in your 3scale installation.
+              - _Application plan approval_: select the check box to require applications getting approval before they can access your API.
+        1. Click <strong>Create application plan</strong>
```
It might be better to have one line between No.114 and No.115

deernano commented on app/javascript/src/QuickStarts/templates/basicAPIintegration.yaml:
```diff
@@ -0,0 +1,200 @@
+        1. On the _New application page_, fill in the details. 
+        
+          Important fields:
+              - _Application plan_: An application is always associated with an application plan.
```
It might be better to have one line between No.138 and No.139

deernano commented on app/javascript/src/QuickStarts/templates/createBackend.yaml:
```diff
@@ -0,0 +1,56 @@
+        1. You have two ways to start the creating process:
+            - Go to the _Dashboard_: Under the APIs section, click <strong>Create backend</strong> in the _Backends_ card. 
+            - Click the context selector on the header and select _Backends_:  On the _Backend_ page, click <strong>Create backend</strong>.
+        1. On the _New backend_ page, fill in the details to create the backend.
```
It is better to have one line between No.35 and No.36

deernano commented on app/javascript/src/QuickStarts/templates/createBackend.yaml:
```diff
@@ -0,0 +1,56 @@
+          Important fields:
+            - _System name_: 3scale needs this name to internally identify and refer to the backend.
+            - _Private base URL_: It is the base URL endpoint of the private API.
+        1. Click <strong>Create backend</strong>
```
It is better to have one line between No.40 and No.41

deernano commented on app/javascript/src/QuickStarts/templates/createBackend.yaml:
```diff
@@ -0,0 +1,56 @@
+      # optional - the task's Check your work module
+      review:
+        instructions: |-
+          - Did you complete the task successfully?
```
Change to  [Did you set the system name for your backend?]

deernano commented on app/javascript/src/QuickStarts/templates/createProduct.yaml:
```diff
@@ -0,0 +1,56 @@
+        1. You have two ways to start the creating process:
+            - Go to the _Dashboard_: Under the APIs section, click <strong>Create product</strong> in the _Products_ card.
+            - Click the context selector on the header and select _Products_: On the _Products_ page, click <strong>Create product</strong>.
+        1. On the _New product_ page, fill in the details to create the product.
```
It is better to have a line between No.35 and No.36

deernano commented on app/javascript/src/QuickStarts/templates/createProduct.yaml:
```diff
@@ -0,0 +1,56 @@
+          Important fields:
+            - _System name_: 3scale needs this name to internally identify and refer to the product.
+            - A product system_name is used to generate proxy endpoints and domain names.
+        1. Click <strong>Create product</strong>
```
It is better to have a line between No.40 and No.41

deernano commented on app/javascript/src/QuickStarts/templates/createApplicationPlan.yaml:
```diff
@@ -0,0 +1,59 @@
+        Application plans define the different sets of access rights you might want to allow for consumers of your API. These can determine anything from rate limits, which methods or resources are accessible and which features are enabled.
+
+        ### Prerequiaites
+        - A product. To create one, see Create new products. 
```
"Create new products" should be a link which link to  createProduct.yaml

dfennessy replied:
@josemigallas  How do I link to the other yaml files?

For example:
- A product. To create one, see [Create new products](./createProduct.yaml)

deernano replied:
Are we able to link to other quick start content within the panel?  For eg. In creating method, there is a link which guide users to 'create backend quick start'.

deernano commented on app/javascript/src/QuickStarts/templates/createApplicationPlan.yaml:
```diff
@@ -0,0 +1,59 @@
+        1. You have two ways to start the creating process:
+            - Go to _[Your\_product\_name]_ overview page: Click <strong>Create application plan</strong>.
+            - Navigate to <strong>[Your\_product\_name] > Applications > Application Plans</strong>: Click <strong>Create application plan</strong>.
+        1. On the _New application plan_ page, fill in the details.
```
It is better to have a line between No.38 and No.39

deernano commented on app/javascript/src/QuickStarts/templates/createApplicationPlan.yaml:
```diff
@@ -0,0 +1,59 @@
+          Important fields:
+            - _System name_: A system name must be unique in your 3scale installation.
+            - _Applications require approval?_: Select the check box to require applications getting approval before they can access your API.
+        1. Click <strong>Create application plan</strong>
```
It is better to have a line between No.43 and No.44

deernano commented on app/javascript/src/QuickStarts/templates/createApplication.yaml:
```diff
@@ -0,0 +1,56 @@
+    - title: Create an application
+      description: |-
+        An application is always associated with an application plan.
+        
```
Prerequisites are missing? Please refer to https://docs.google.com/document/d/1r9rnutZuF4j9_zRZl1MIBya95vlz4pFuhxiNB8QGBtE/edit#v

deernano commented on app/javascript/src/QuickStarts/templates/createApplication.yaml:
```diff
@@ -0,0 +1,56 @@
+
+          Important fields:
+            - _Application plan_: An application is always associated with an application plan. If there is no existing application plan, click <strong>Create a new application plan</strong>.
+        1. Click <strong>Create application</strong>
```
It is better to have a line between No.40 and No.41

deernano commented on app/javascript/src/QuickStarts/templates/createMethod.yaml:
```diff
@@ -0,0 +1,101 @@
+
+        ### Prerequisites
+
+        - A product. To create one, see Create a product.
```
see" Create a product" is a link which link to ' createProduct.yaml'

deernano commented on app/javascript/src/QuickStarts/templates/createMethod.yaml:
```diff
@@ -0,0 +1,101 @@
+
+        ### Prerequisites
+
+        - A backend. To create one, see Create a backend.
```
see" Create a backend" is a link which link to ' createBackend.yaml'

deernano commented on app/javascript/src/QuickStarts/templates/createMethod.yaml:
```diff
@@ -0,0 +1,101 @@
+        
+          Important fields:
+              - _Friendly name_: A short description of the method. This name is displayed in different sections of the 3scale Admin Portal.
+              - _System name_: The name of the method in your API to use to report the usage through the 3scale Service Management API. The system name must be unique.
```
It is better to have a line between No.48 and No.49

deernano commented on app/javascript/src/QuickStarts/templates/createMethod.yaml:
```diff
@@ -0,0 +1,101 @@
+        ### Prerequisites
+
+        - A product. To create one, see Create a product.
+        - An application plan. To create one, see Create an application plan.
```
see "Create an application plan" is a link, which link to 'createApplicationPlan.yaml'

deernano commented on app/javascript/src/QuickStarts/templates/createMethod.yaml:
```diff
@@ -0,0 +1,101 @@
+          Important fields:
+              - _Friendly name_: A short description of the method. This name is displayed in different sections of the 3scale Admin Portal.
+              - _System name_: The name of the method in your API to use to report the usage through the 3scale Service Management API. The system name must be unique.
+        1. Click <strong>Create method</strong>
```
It is better to have a line between No.78 and No.79

deernano commented on app/javascript/src/QuickStarts/templates/createMetric.yaml:
```diff
@@ -0,0 +1,106 @@
+
+        ### Prerequisites
+
+        - A backend. To create one, see Creating a backend.
```
see 'Creating a backend' is a link, which link to 'createBackend.yaml'

deernano commented on app/javascript/src/QuickStarts/templates/createMetric.yaml:
```diff
@@ -0,0 +1,106 @@
+              - _Friendly name_: A short description of the metric. This name is displayed in different sections of the 3scale Admin Portal.
+              - _System name_: The name of the metric in your API to use to report the usage through the 3scale Service Management API. The system name must be unique.
+              
+              > <strong>Tip:</strong> Be careful with changing the system name of the metrics or deleting them. These changes can break your already deployed 3scale integration if there are mapping rules pointing to the previous system name of the metric.
```
It is better to have a line between No.49 and No.50

deernano commented on app/javascript/src/QuickStarts/templates/createMetric.yaml:
```diff
@@ -0,0 +1,106 @@
+
+        ### Prerequisites
+
+        - A Product. To create one, see Creating a product.
```
see 'Creating a product' is a link, which link to 'createProduct.yaml'

deernano commented on app/javascript/src/QuickStarts/templates/createMetric.yaml:
```diff
@@ -0,0 +1,106 @@
+        ### Prerequisites
+
+        - A Product. To create one, see Creating a product.
+        - An application plan. To create one, see Create an application plan.
```
see 'Create an application plan' is a link, which link to 'createApplicationPlan.yaml'

deernano commented on app/javascript/src/QuickStarts/templates/createMetric.yaml:
```diff
@@ -0,0 +1,106 @@
+              - _Friendly name_: A short description of the metric. This name is displayed in different sections of the 3scale Admin Portal.
+              - _System name_: The name of the metric in your API to use to report the usage through the 3scale Service Management API. The system name must be unique.
+              - _Unit_: Use a singular noun, for example, _hit_. The singular will become plural in the analytics charts.
+        1. Click <strong>Create metric</strong>
```
It is better to have a line between No.84 and No.85

deernano commented on app/javascript/src/QuickStarts/templates/createMappingRule.yaml:
```diff
@@ -0,0 +1,126 @@
+
+        ### Prerequisites
+
+        - A backend. To create one, see Create a backend.
```
see "Create a backend"  is a link ,which link to 'createBackend.yaml'

deernano commented on app/javascript/src/QuickStarts/templates/createMappingRule.yaml:
```diff
@@ -0,0 +1,126 @@
+        ### Prerequisites
+
+        - A backend. To create one, see Create a backend.
+        - A method. Mapping rules are required in your previously created methods, however, they are optional for metrics, To create one, see Create a method.
```
'Create a method' is a link, which link to "createMethod.yaml'

dfennessy replied:
I've created a link for this one, but I cannot preview it. It shows an error for me. Maybe there is another way to test it other than clicking it from the preview to see if it loads the content for Create a method in the createMethod.yaml file.

deernano commented on app/javascript/src/QuickStarts/templates/createMappingRule.yaml:
```diff
@@ -0,0 +1,126 @@
+              - _Last?_: If this mapping rule should be considered as the last one, to stop processing other mapping rules.
+              - _Position_: Number that indicates the position of the execution of the mapping rule, to sort the mapping rules.
+
+        1. Click <strong>Create mapping rule</strong>
```
It is better to have a line between No.64 and No.65

dfennessy commented:
Thanks for the comment @deernano 

I need further guidance on how to remove the whitespace in the lists. Either a markdown solution or css maybe.

I also need guidance on how to add links for the content between Yaml file.

Hopefully @josemigallas can help with this.

Cheers.

dfennessy commented:
Myself and @josemigallas have a meeting to discuss the links this afternoon.

josemigallas commented on app/javascript/src/QuickStarts/templates/links.js:
```diff
@@ -0,0 +1,5 @@
+// @flow
+
+export default [
+  ['this-text-from-the-yml-will-be-replaced', '/with/this/href', 'And this text in the HTML']
```
1. The exact text that will be replaced from the yaml. It will replace the first match, therefore it must be a unique identifier that can't be accidentally reused.
2. The URL used in the anchor tag.
3. The text the user will see in the guide.

dfennessy commented on app/javascript/src/QuickStarts/templates/createApplication.yaml:
```diff
@@ -0,0 +1,60 @@
+# Additional info: https://docs.openshift.com/container-platform/4.9/web_console/creating-quick-start-tutorials.html
+metadata:
+  name: getting-started-with-quick-starts
```
Updating these to be unique.

josemigallas commented on app/presenters/quickstarts_presenter.rb:
```diff
@@ -0,0 +1,22 @@
+
+  def links
+    [
+      ['create-a-product-link', new_admin_service_path, 'Create a product'],
+      ['create-a-backend-link', new_provider_admin_backend_api_path, 'Create a backend'],
+      ['create-an-application-plan-link', new_admin_service_application_plan_path(current_api), 'Create an application plan'],
+      ['create-a-method-link-in-create-mapping-rules', new_admin_service_metric_child_path(current_api, current_api.metrics.hits), 'Create a method']
```
@dfennessy We're gonna do it this way in order to have dynamic links. We rely on `current_api` which is a Product or Backend either:
- the one in the scope (for instance if we're at the Overview page of a product/backend)
- or the default service, which is I guess the first API (product)

dfennessy commented:
Hi @junezhang @deernano @josemigallas

Can this PR be merged soon?

deernano commented:
> 

Yes,  it is able to merge

dfennessy commented:
> > 
> 
> Yes, it is able to merge

Thanks, Melody. 

@josemigallas Please merge :smile:
